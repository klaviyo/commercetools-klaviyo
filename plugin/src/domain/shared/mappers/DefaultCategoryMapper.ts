import { Category } from '@commercetools/platform-sdk';
import { CategoryMapper } from './CategoryMapper';
import { getLocalizedStringAsText } from '../../../utils/locale-currency-utils';

export class DefaultCategoryMapper implements CategoryMapper {
    public mapCtCategoryToKlaviyoCategory(category: Category, klaviyoCategoryId?: string): CategoryRequest {
        return {
            data: {
                type: 'catalog-category',
                id: klaviyoCategoryId,
                attributes: {
                    external_id: !klaviyoCategoryId ? category.id : undefined,
                    name: this.getCategoryNameWithAncestors(category),
                    integration_type: !klaviyoCategoryId ? '$custom' : undefined,
                    catalog_type: !klaviyoCategoryId ? '$default' : undefined,
                },
            },
        };
    }

    private getCategoryNameWithAncestors(category: Category): string {
        const categoryNames = category.ancestors
            .map((ancestor) => getLocalizedStringAsText((ancestor.obj as Category).name))
            .concat(getLocalizedStringAsText(category.name));
        return categoryNames.join(' > ');
    }

    public mapKlaviyoCategoryIdToDeleteCategoryRequest(klaviyoCategoryId: string): CategoryDeletedRequest {
        return {
            data: {
                id: klaviyoCategoryId,
            },
        };
    }
}
