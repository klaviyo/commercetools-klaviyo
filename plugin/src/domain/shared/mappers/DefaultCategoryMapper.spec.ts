import { DefaultCategoryMapper } from './DefaultCategoryMapper';
import { sampleCategoryCreatedMessage } from '../../../test/testData/ctCategoryMessages';

const categoryMapper = new DefaultCategoryMapper();

describe('map CT category to Klaviyo category', () => {
    it('should map a commercetools category to a klaviyo category', () => {
        const klaviyoEvent = categoryMapper.mapCtCategoryToKlaviyoCategory(sampleCategoryCreatedMessage.category);
        expect(klaviyoEvent).toMatchSnapshot();
    });

    it('should extend the klaviyo category name if the category is a subcategory (has ancestors)', () => {
        const klaviyoEvent = categoryMapper.mapCtCategoryToKlaviyoCategory({
            ...sampleCategoryCreatedMessage.category,
            ancestors: [{
                typeId: 'category',
                id: '123456',
                obj: {
                    name: {
                        'en-US': 'Test main category',
                    },
                } as any,
            }, 
            {
                typeId: 'category',
                id: '123456',
                obj: {
                    name: {
                        'en-US': 'Test subcategory',
                    },
                } as any,
            }],
        });
        expect(klaviyoEvent).toMatchSnapshot();
    });

    it('should map a commercetools category to a klaviyo category with an existing category id', () => {
        const klaviyoEvent = categoryMapper.mapCtCategoryToKlaviyoCategory(
            sampleCategoryCreatedMessage.category,
            '123456',
        );
        expect(klaviyoEvent).toMatchSnapshot();
    });
});

describe('map Klaviyo category to Klaviyo delete category request', () => {
    it('should map a klaviyo category id to a klaviyo delete category', () => {
        const klaviyoEvent = categoryMapper.mapKlaviyoCategoryIdToDeleteCategoryRequest(sampleCategoryCreatedMessage.category.id);
        expect(klaviyoEvent).toMatchSnapshot();
    });
});
