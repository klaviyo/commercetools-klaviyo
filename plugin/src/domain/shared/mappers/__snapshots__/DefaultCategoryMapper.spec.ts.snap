// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map CT category to Klaviyo category should extend the klaviyo category name if the category is a subcategory (has ancestors) 1`] = `
{
  "data": {
    "attributes": {
      "catalog_type": "$default",
      "external_id": "cd2cd62b-7f08-44e4-96a8-d3e34c40cac9",
      "integration_type": "$custom",
      "name": "Test main category > Test subcategory > MessageTriggerTest1",
    },
    "id": undefined,
    "type": "catalog-category",
  },
}
`;

exports[`map CT category to Klaviyo category should map a commercetools category to a klaviyo category 1`] = `
{
  "data": {
    "attributes": {
      "catalog_type": "$default",
      "external_id": "cd2cd62b-7f08-44e4-96a8-d3e34c40cac9",
      "integration_type": "$custom",
      "name": "MessageTriggerTest1",
    },
    "id": undefined,
    "type": "catalog-category",
  },
}
`;

exports[`map CT category to Klaviyo category should map a commercetools category to a klaviyo category with an existing category id 1`] = `
{
  "data": {
    "attributes": {
      "catalog_type": undefined,
      "external_id": undefined,
      "integration_type": undefined,
      "name": "MessageTriggerTest1",
    },
    "id": "123456",
    "type": "catalog-category",
  },
}
`;

exports[`map Klaviyo category to Klaviyo delete category request should map a klaviyo category id to a klaviyo delete category 1`] = `
{
  "data": {
    "id": "cd2cd62b-7f08-44e4-96a8-d3e34c40cac9",
  },
}
`;
